infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/logreader/StreamsSnapshotReader.java

```diff
@@ -94,23 +100,35 @@ private OpaqueEntry generateOpaqueEntry(long version, UUID streamID,  SMREntryLi
      * @param entryList
      * @return
      */
-    private LogReplicationEntry generateMessage(OpaqueStreamIterator stream, SMREntryList entryList, UUID snapshotRequestId) {
+    private LogReplicationEntryMsg generateMessage(
+            OpaqueStreamIterator stream,
+            SMREntryList entryList,
+            UUID snapshotRequestId) {
         currentMsgTs = stream.maxVersion;
         OpaqueEntry opaqueEntry = generateOpaqueEntry(currentMsgTs, stream.uuid, entryList);
         if (!stream.iterator.hasNext()) {
             //mark the end of the current stream.
             currentMsgTs = snapshotTimestamp;
         }
 
-        LogReplicationEntry txMsg = new LogReplicationEntry(MessageType.SNAPSHOT_MESSAGE, topologyConfigId, snapshotRequestId, currentMsgTs,
-                preMsgTs, snapshotTimestamp, sequence, opaqueEntry);
+        LogReplication.LogReplicationEntryMetadataMsg metadata = LogReplication.LogReplicationEntryMetadataMsg.newBuilder()
+                .setEntryType(LogReplicationEntryType.SNAPSHOT_MESSAGE)
+                .setTopologyConfigID(topologyConfigId)
+                .setSyncRequestId(getUuidMsg(snapshotRequestId))
+                .setTimestamp(currentMsgTs)
+                .setPreviousTimestamp(preMsgTs)
+                .setSnapshotTimestamp(snapshotTimestamp)
+                .setSnapshotSyncSeqNum(sequence)
+                .build();
+
+        LogReplicationEntryMsg txMsg = getLrEntryMsg(ByteString.copyFrom(generatePayload(opaqueEntry)), metadata);
```

Comment:
Since `generatePayload` returns the backing byte array of the `ByteBuf` used, it could be more efficient to copy from `readerIndex()` to `writeIndex()` since the backing array can be larger than the amount of data written.

Reply:
Looking at the API, we can use `unsafeWrap`. There is no sharing of the ownership.